parmed/tools/actions.py

```diff
@@ -3243,13 +3246,62 @@ class HMassRepartition(Action):
     def init(self, arg_list):
         self.changewater = arg_list.has_key('dowater')
         self.new_h_mass = arg_list.get_next_float(optional=True, default=3.024)
+        self.alt_new_h_mass = arg_list.get_key_float('altmass', default=None)
 
     def __str__(self):
         retstr = f'Repartitioning hydrogen masses to {self.new_h_mass} daltons. '
         if self.changewater:
             return retstr + 'Also changing water hydrogen masses.'
         return retstr + 'Not changing water hydrogen masses.'
 
+    def _residue_to_rdkit(self, residue):
+        """Extract a residue as a standalone rdkit Mol object."""
+        # Hack to make sure that atom types are compliant with rdkit
+        # expectations. That is, rdkit must be able to infer the element based
+        # solely on the atom type, which is not always true for things like
+        # standard GAFF types.
+        #
+        mol = self.parm[':%d'%(residue.number + 1)]
+        for atom in mol.atoms:
+            atom.type = element_by_name(atom.name)
+
+        rslt = io.StringIO()
+        Mol2File.write(mol, rslt)
+        mol2str = rslt.getvalue()
+        # clean the rslt string
+        rslt.truncate(0)
+        rslt.seek(0)
+        rmol = Chem.MolFromMol2Block(mol2str, sanitize=False)
+        return rmol
+
+    def is_in_ring(self, atom):
+        # The heavy atom must have atleast two bonds with other heavy atoms and
+        # no more than two bonds with hydrogen atoms (assuming a max
+        # coordination of four). This quickly avoids silly cases like
+        # accidentally flagging a rigid water as a three-membered ring.
+        #
+        num_hyd_partners =\
+                sum((bp.atomic_number == 1) for bp in atom.bond_partners)
+        num_hvy_partners = len(atom.bond_partners) - num_hyd_partners
+        if not (num_hvy_partners >= 2 and num_hyd_partners in [1, 2]):
+            return False
+        # Use rdkit Atom.IsInRing() method.
+        # This is obviously really inefficient since each atom in a residue
+        # will cause the _same_ molecule object to be built. It doesn't seem to
+        # be a terrible speed issue for now and it can always be restructured
+        # later.
+        #
+        rmol = self._residue_to_rdkit(atom.residue)
```

Comment:
you could `lru_cache` this (after adding a `__hash__` implementation to `class Residue` that simply returns `id(self)`).

This algorithm is O(N^2) in # of atoms, but adding the cache would drop this to O(N).

Reply:
Lookup table is already there.